lib/spack/spack/build_systems/cargo.py

```diff
@@ -0,0 +1,247 @@
+
+        target_cpu = self.target_cpu
+        if target_cpu:
+            rustflags.extend(["-C", "target-cpu={0}".format(target_cpu)])
+
+        # Put config in spack-src so that it does not overwrite the config
+        # from vendoring dependencies
+        mkdirp(join_path(self.stage.source_path, '.cargo'))
+        config_path = join_path(self.stage.source_path, '.cargo/config')
+
+        build_type = self.spec.variants['build_type'].value
+        lto = self.spec.variants['lto'].value
+
+        profile = 'dev' if build_type == 'debug' else build_type
+        lto = 'off' if lto == 'none' else lto
+
+        with open(config_path, 'w') as f:
+            f.write("""
+[build]
+target = "{target}"
+rustflags = "{flags}"
+
+[profile.{profile}]
+rpath = true
+lto = "{lto}"
+""".format(
+                target=target,
+                flags=" ".join(rustflags),
+                profile=profile,
+                lto=lto
+            ))
+
+    def workspace_targets(self):
+        metadata = self.metadata
+        for p in metadata["packages"]:
+            for w in metadata["workspace_members"]:
+                if p["id"] == w:
+                    for t in p["targets"]:
+                        yield t
+
+    def cargo_features(self):
+        """Features to build cargo project with. Overridable by package."""
+        return None
+
+    def _feature_args(self):
+        """Returns the args associated with cargo_features"""
+        features = self.cargo_features()
+        if features is None:
+            return []
+
+        args = ["--no-default-features"]
+
+        for feature in features:
+            args += ["--features", feature]
+
+        return args
+
+    def build(self, spec, prefix):
+        """cargo build"""
+        jobs = spack.config.get('config:build_jobs') if self.parallel else 1
+
+        # Explicitly build each target in the workspace
+        for t in self.workspace_targets():
+            args = [
+                "rustc", "--jobs", str(jobs), "-vv", "--manifest-path",
+                self.manifest_path
+            ]
+
+            if 'build_type=release' in self.spec:
+                args += ["--release"]
+
+            args += self._feature_args()
+
+            if "bin" in t["kind"]:
+                args += ["--bin"]
+            elif "staticlib" in t["kind"]:
+                args += ["--lib"]
+            elif "cdylib" in t["kind"]:
+                args += ["--lib"]
+            else:
+                continue
+
+            args += [t["name"], "--"]
+
+            args += ["--cap-lints", "warn"]
+            if '+prefer_dynamic' in self.spec:
+                args += ["--codegen", "prefer-dynamic"]
+
+            self.cargo(*args)
+
+    def install(self, spec, prefix):
+        """Install cargo targets"""
+
+        # This is not implemented using "cargo install" because we want to
+        # install libraries, too.
+        so_ext = 'dylib' if 'platform=darwin' in self.spec else 'so'
```

Comment:
This should already be defined elsewhere in Spack

Reply:
So there's [dso_suffix](https://github.com/spack/spack/blob/develop/lib/spack/spack/build_environment.py#L96). However, I prefer my version because it's derived from the spec's platform, not from the host's platform. It probably doesn't matter since spack doesn't really currently support cross compiling between platforms all that well, but maybe someday?